### ISSUE TITLE
unable to `kubectl patch` change to a secret name

### ISSUE THREAD

**Issue by iameli** [kubernetes/kubernetes#14447]
Hello,

I'd like to switch a secret being used in the podspec of a replicationcontroller using `kubectl patch`. The error:

> Error from server: unable to find api field in struct VolumeSource for the json field "secretName"

Mayhaps I'm doing something wrong, and it's just a cryptic error message as referenced in #3338?

**command output**

```
> kubectl create -f nginx-terminator.rc.json 
replicationcontrollers/nginx-terminator

> kubectl patch rc nginx-terminator -p "`cat patch.json`"
Error from server: unable to find api field in struct VolumeSource for the json field "secretName"
```

**nginx-terminator.rc.json**
```
{
  "kind": "ReplicationController",
  "apiVersion": "v1",
  "metadata": {
    "name": "nginx-terminator",
    "namespace": "csats-pizza"
  },
  "spec": {
    "replicas": 2,
    "selector": {
      "app": "nginx-terminator",
      "version": "1"
    },
    "template": {
      "metadata": {
        "labels": {
          "app": "nginx-terminator",
          "version": "1"
        }
      },
      "spec": {
        "volumes": [
          {
            "name": "ssl-keys",
            "secret": {
              "secretName": "wildcard-ssl"
            }
          }
        ],
        "containers": [
          {
            "name": "nginx-terminator",
            "image": "registry.csats.pizza/csats/nginx-terminator",
            "ports": [
              {
                "containerPort": 443,
                "protocol": "TCP"
              }
            ],
            "imagePullPolicy": "Always",
            "volumeMounts": [
              {
                "name": "ssl-keys",
                "readOnly": true,
                "mountPath": "/certs"
              }
            ]
          }
        ],
        "restartPolicy": "Always",
        "dnsPolicy": "ClusterFirst"
      }
    }
  }
}
```

**patch.json**
```
{
  "spec": {
    "template": {
      "spec": {
        "volumes": [
          {
            "name": "ssl-keys",
            "secret": {
              "secretName": "wildcard-ssl-2"
            }
          }
        ]
      }
    }
  }
}
```

```
> kubectl version
Client Version: version.Info{Major:"1", Minor:"0", GitVersion:"v1.0.6", GitCommit:"388061f00f0d9e4d641f9ed4971c775e1654579d", GitTreeState:"clean"}
Server Version: version.Info{Major:"1", Minor:"0", GitVersion:"v1.0.6", GitCommit:"388061f00f0d9e4d641f9ed4971c775e1654579d", GitTreeState:"clean"}
```

**Comment by erictune** [comment]
We'll take a look at this.  In the meantime, you can delete the rc with  `kubectl delete --cascade=false ...` and make a new one.  The pods will keep running and the new one will adopt them.

**Comment by iameli** [comment]
Thanks! I ended up doing the opposite -- just changed the secret and restarted the pods.

**Comment by bgrant0607** [comment]
VolumeSource is an inline field:
https://github.com/kubernetes/kubernetes/blob/master/pkg/api/v1/types.go#L198

@jackgr @ghodss Looks like a bug in strategic merge patch.

**Comment by bgrant0607** [comment]
Should have been fixed by #14985
